### PR TITLE
Allow error, object, xml module prefixes without importing

### LINKED ISSUE #23099: Can refer to error, object, xml module prefix without importing
**Description:**
Align auto imports with the spec.

**8.1 Import declaration** says:
> 
A module prefix of t, where t is one of error, object or xml, is predeclared as referring to the lang.t lang library module, but this can be overridden by an import-decl. 

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by chiranSachintha**
> 8.1 Import declaration says:
 A module prefix of t, where t is one of error, object or xml, is predeclared as referring to the lang.t lang library module, but this can be overridden by an import-decl.

Fixes #23099 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java [hunk 1]
```diff
@@ -185,6 +185,7 @@
 
     private Names names;
     public Map<BPackageSymbol, SymbolEnv> pkgEnvMap = new HashMap<>();
+    public Map<Name, BPackageSymbol> predeclaredModules = new HashMap<>();
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java:188]: A minor thing. Was wondering whether we should make this an immutable map.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 2]
```diff
@@ -533,6 +535,25 @@ public void visit(BLangImportPackage importPkgNode) {
         this.env.scope.define(pkgAlias, symbol);
     }
 
+    public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules, SymbolEnv env) {
+        SymbolEnv prevEnv = this.env;
+        this.env = env;
+        for (Name alias : predeclaredModules.keySet()) {
+            ScopeEntry entry = this.env.scope.lookup(alias);
+            if (predeclaredModules.get(alias) != null) {
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:543]: This null check is not necessary right? We are iterating through the key set of the predeclared modules map. Therefore, there's no reason for the map to not contain the key it just retrieved

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Fixed

File: tests/jballerina-unit-test/src/test/resources/test-src/imports/predeclared-imports/src/bar/predeclared-error.bal [hunk 3]
```diff
@@ -0,0 +1,4 @@
+
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/imports/predeclared-imports/src/bar/predeclared-error.bal:16]: Done

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 4]
```diff
@@ -528,6 +530,25 @@ public void visit(BLangImportPackage importPkgNode) {
...
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: With the changes this logic removed

[the remark this answers is not in the document]
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: With the changes this logic removed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 5]
```diff
@@ -395,6 +395,9 @@ public BSymbol resolvePrefixSymbol(SymbolEnv env, Name pkgAlias, Name compUnit)
             }
 
             if ((entry.symbol.tag & SymTag.IMPORT) == SymTag.IMPORT &&
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java:397]: Shall we refactor this into something like

```java
            if ((entry.symbol.tag & SymTag.IMPORT) == SymTag.IMPORT) {
                Name importCompUnit = ((BPackageSymbol) entry.symbol).compUnit;
                if (importCompUnit == null) {
                    return entry.symbol;
                } else if (importCompUnit.equals(compUnit)) {
                    ((BPackageSymbol) entry.symbol).isUsed = true;
                    return entry.symbol;
                }
            }
```

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java:397]: Let's also add a comment to say why the compUnit can be null here?

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Done

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/imports/ImportsNegativeTests.java [hunk 6]
```diff
@@ -39,4 +39,16 @@ public void testInvalidImportOnMultipleFiles() {
...
+    @Test(description = "Test invalid autoimports")
+    public void testInvalidAutoImports() {
+        CompileResult result = BCompileUtil.compile("test-src/imports/predeclared-imports", "rpg");
+        assertEquals(result.getErrorCount(), 5);
+        validateError(result, 0, "undefined function 'max'", 3, 12);
+        validateError(result, 1, "undefined function 'min'", 7, 12);
+        validateError(result, 2, "undefined function 'concat'", 8, 12);
+        validateError(result, 3, "unknown type 'Listener'", 12, 6);
+        validateError(result, 4, "unknown type 'CallStackElement'", 35, 30);
+
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/imports/ImportsNegativeTests.java]: Fixed

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/definitions/TestablePackageCompletionTest.java [hunk 7]
```diff
@@ -35,7 +35,7 @@
         log.info("Test textDocument/completion for Function Definition Scope");
         return new Object[][] {
 //                {"topLevelCompletion.json", "testablepackage"},
-                {"blockStatementCompletion.json", "testablepackage"},
+//                {"blockStatementCompletion.json", "testablepackage"},
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/definitions/TestablePackageCompletionTest.java]: No, I will create

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java (hunk 1 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java:191]: Changed to immutable map.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java [hunk 8]
```diff
@@ -347,6 +349,14 @@ public BType getLangLibSubType(String name) {
         throw new IllegalStateException("LangLib Subtype not found: " + name);
     }
 
+    public void loadPredeclaredModules() {
+        this.predeclaredModules.put(Names.ERROR, this.langErrorModuleSymbol);
+        this.predeclaredModules.put(Names.OBJECT, this.langObjectModuleSymbol);
+        this.predeclaredModules.put(Names.XML, this.langXmlModuleSymbol);
+
+        this.predeclaredModules = Collections.unmodifiableMap(this.predeclaredModules);
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java:357]: What I meant was something like
```java
Map<Name, BPackageSymbol> modules = new HashMap<>();
modules.put(Names.ERROR, this.langErrorModuleSymbol);
modules.put(Names.OBJECT, this.langObjectModuleSymbol);
modules.put(Names.XML, this.langXmlModuleSymbol);

this.predeclaredModules = Collections.unmodifiableMap(modules);
```

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/definitions/TestablePackageCompletionTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/definitions/TestablePackageCompletionTest.java]: https://github.com/ballerina-platform/ballerina-lang/issues/24153

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java (hunk 8 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java:357]: It gives NullPointerException while building blang packages in this way.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/imports/ImportsTest.java [hunk 9]
```diff
@@ -63,4 +63,10 @@ public void testSameModuleNameDifferentOrgImports() {
         BValue[] returns = BRunUtil.invoke(result, "getStringValueOfPI");
         Assert.assertTrue((returns[0]).stringValue().startsWith("3.14"));
     }
+
+    @Test(description = "Test auto imports")
+    public void testPredeclaredModules() {
+        BCompileUtil.compile("test-src/imports/predeclared-imports", "bar");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/imports/ImportsTest.java:69]: Are we only building these or are we also validating the functionality?

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/imports/ImportsTest.java:69]: Only building that module. Is it need to validate functionality here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java (hunk 8 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java]: Changed (As mention in the https://github.com/ballerina-platform/ballerina-lang/pull/23884#discussion_r440125692)

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/imports/ImportsTest.java (hunk 9 above)
Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/imports/ImportsTest.java:69]: Could you please send a separate PR validating 0 diagnostics.

**Pull request merged**